### PR TITLE
Always include the last-resort font

### PR DESCRIPTION
**Pull request by mbrubeck**
This is used as a fallback for any characters that don't have glyphs in the specified font.  Without this, per-glyph font fallback doesn't work because the FontGroup always contains only one font.  Fixes missing glyphs on many pages on my Linux box.

As a follow-up, we should probably have a smarter strategy for finding fallback fonts, possibly varying by script.  (Currently we just have a few hard-coded family names.)

r? @glennw

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11303)

### CODE REVIEW

**Comment by gsnedders** [comment]
This hopefully fixes (I believe currently disabled) #7625.

**Comment by mbrubeck** [comment]
>   ▶ FAIL [expected PASS] /css-text-3_dev/html/word-break-break-all-007.htm
>   ▶ FAIL [expected PASS] /css-text-3_dev/html/word-break-normal-hi-000.htm

I think these were passing accidentally before; we are (and were) rendering them with fonts that don’t contain the required glyphs.

>   ▶ FAIL [expected PASS] /css21_dev/html4/block-in-inline-003.htm
>   ▶ FAIL [expected PASS] /css21_dev/html4/block-in-inline-008.htm
>   ▶ FAIL [expected PASS] /css21_dev/html4/list-style-applies-to-008.htm
>   ▶ FAIL [expected PASS] /css21_dev/html4/list-style-type-applies-to-008.htm

These all look like they’re affected by the same vertical positioning bug, like maybe an empty line was inserted where it shouldn’t be.

Also, mac-rel-wpt:
```
Ran 4838 tests finished in 1012.0 seconds.
  • 4837 ran as expected. 894 tests skipped.
  • 1 tests failed unexpectedly

Tests with unexpected results:
  ▶ FAIL [expected PASS] /_mozilla/css/per_glyph_font_fallback_a.html
```
Looking at the reftest analyzer, the glyph fallback is working correctly but the line height is different in the test versus the reference; the patch must have changed which font metrics are being used.

On the Linux builder there were even more affected tests:
* test-css: 58 tests failed unexpectedly and 7 tests passed unexpectedly
* test-wpt: 6 tests failed unexpectedly

All of these seem to be affected by the same line-height issue, except `/_mozilla/css/pre_with_tab.html` which has an issue with horizontal spacing instead.  I can reproduce the Linux failures locally.  I suspect all the behavior changes have the same root cause: some sort of change in which font metrics are used when fallback is triggered.

**Comment by mbrubeck** [comment]
Also, [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1139):
```
  ▶ FAIL [expected PASS] /_mozilla/css/per_glyph_font_fallback_a.html
```
and [linux-rel-css](http://build.servo.org/builders/linux-rel/builds/1141):
```
Ran 9969 tests finished in 952.0 seconds.
  • 9721 ran as expected. 129 tests skipped.
  • 13 tests failed unexpectedly
  • 235 tests passed unexpectedly
```

**Comment by mbrubeck** [comment]
@bors-servo try

* Now with over 250 tests newly marked passing!

**Comment by mbrubeck** [comment]
Looks like there are still 13 css21_dev tests failing on Linux only, all because of vertical alignment spacing issues involving Ahem.ttf.

**Comment by mbrubeck** [comment]
The latest push fixes those last few test failures.  They were caused by layout using metrics from the fallback font instead of the specified font.

r? @glennw or @pcwalton.

**Comment by mbrubeck** [comment]
Sigh.  The fix for the Ahem metrics on Linux broke `{fonts,font-family}-013.htm` on both platforms.  The font fallback is working correctly but now the vertical metrics are wrong.  These tests were disabled previously because of #7625, so I just marked them as failing for now.

File: components/layout/fragment.rs [hunk 1]
```diff
@@ -1970,11 +1970,14 @@ impl Fragment {
...
-            SpecificFragmentInfo::ScannedText(ref text_fragment) => {
+            SpecificFragmentInfo::ScannedText(_) => {
                 // See CSS 2.1 § 10.8.1.
                 let line_height = self.calculate_line_height(layout_context);
+                let font_style = self.style.get_font_arc();
+                let font_metrics = text::font_metrics_for_style(&mut layout_context.font_context(),
+                                                                font_style);
                 let font_derived_metrics =
-                    InlineMetrics::from_font_metrics(&text_fragment.run.font_metrics, line_height);
+                    InlineMetrics::from_font_metrics(&font_metrics, line_height);
```
Comment (pcwalton) [inline components/layout/fragment.rs]: Why this change?

Comment (mbrubeck) [inline components/layout/fragment.rs]: Without this, several tests that use Ahem had incorrect vertical placement of lines.  However, reading the spec again it seems this change is incorrect, since it should use the actual font for each glyph. I'll look further into why the tests were failing; maybe it has to do with the font chosen for struts...

Comment (mbrubeck) [inline components/layout/fragment.rs]: The issue is caused by including metrics for fragments that consist of only discardable or zero-width characters, like bidi control characters.  I'll work on fixing this next week.

**Comment by mbrubeck** [comment]
@bors-servo try

Fixed the failures in the `-013` tests and removed some incorrect code, by correctly ignoring fragments with no glyphs. `fonts-012.htm` is the only failing test now.  The test file actually renders correctly, but the reference file has a vertical alignment problem. This was passing by accident before because of missing font fallback, so I'd like to mark it failing for now and file a followup to investigate it.

This is ready for re-review assuming no more surprises turn up on the try run.

**Comment by mbrubeck** [comment]
>  ▶ FAIL [expected PASS] /css-color-3_dev/html4/t422-rgba-onscreen-multiple-boxes-c.htm
>  ▶ FAIL [expected PASS] /css-color-3_dev/html4/t425-hsla-onscreen-multiple-boxes-c.htm

These were passing with an incorrect rendering before. Marked as failing.

>  ▶ PASS [expected FAIL] /css21_dev/html4/fonts-012.htm

Marked as expected PASS on mac.

**Comment by pcwalton** [comment]
@bors-servo: r+

Let's go!

**Comment by larsbergstrom** [comment]
@bors-servo r+
 - infra

**Comment by mbrubeck** [comment]
@bors-servo r=pcwalton

* `per_glyph_font_fallback_a.html` is a broken test (failing in Gecko too). Pushed a better fix for it.

**Comment by mbrubeck** [comment]
@bors-servo r=pcwalton

* Pushed a better fix for the broken test, actually tested on Mac locally.
